### PR TITLE
Switch to version Dagger 0.18.11

### PR DESCRIPTION
**Pull request by SmalRat**
BEGINRELEASENOTES
- Switched to version 0.18.11
- Added relevant examples and documentation
- Small fixes

ENDRELEASENOTES

### CODE REVIEW

**Comment by m-fila** [comment]
Should we try to merge it now or wait for https://github.com/JuliaParallel/Dagger.jl/pull/531?
If we do it now we'll get gantt plott but spoil execution graph from logs, right?

**Comment by SmalRat** [comment]
It contains a modified version (with fixes) of GraphVizSimpleExt, which is used by default, so everything works as of 0.18.11.

**Comment by m-fila** [comment]
I see, the `GraphVizSimpleExt` has all the changes from https://github.com/JuliaParallel/Dagger.jl/pull/531 so we can drop it later once the fix gets into dagger

**Comment by m-fila** [comment]
Running `graphs_scheduling/src/main.jl` gives the following error once all the processors are done:
```jl
ERROR: LoadError: KeyError: key 68 not found
Stacktrace:
  [1] getindex
    @ ./dict.jl:498 [inlined]
  [2] _proc_color(ctx::@NamedTuple{proc_to_color::Dict{Dagger.Processor, String}, proc_colors::Vector{RGB{FixedPointNumbers.N0f8}}, proc_color_idx::Base.RefValue{Int64}, proc_to_shape::Dict{Type, String}, proc_shapes::Tuple{String, String, String}, proc_shape_idx::Base.RefValue{Int64}, id_to_proc::Dict{Int64, Dagger.Processor}}, id::Int64)
    @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:176
  [3] write_edge(io::IOStream, ts_move::TimespanLogging.Timespan, logs::Vector{TimespanLogging.Timespan}, ctx::@NamedTuple{proc_to_color::Dict{Dagger.Processor, String}, proc_colors::Vector{RGB{FixedPointNumbers.N0f8}}, proc_color_idx::Base.RefValue{Int64}, proc_to_shape::Dict{Type, String}, proc_shapes::Tuple{String, String, String}, proc_shape_idx::Base.RefValue{Int64}, id_to_proc::Dict{Int64, Dagger.Processor}}, inputname::String, inputarg::RemoteChannel{Channel{Int64}})
    @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:235
  [4] write_dag(io::IOStream, logs::Vector{TimespanLogging.Timespan}, t::Nothing)
    @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:323
  [5] write_dag
    @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:260 [inlined]
  [6] _show_plan
    @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:353 [inlined]
  [7] show_logs
    @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:369 [inlined]
  [8] show_logs(io::IOStream, logs::Vector{TimespanLogging.Timespan}, vizmode::Symbol; options::@Kwargs{})
    @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:363
  [9] show_logs
    @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:363 [inlined]
 [10] #17
    @ ~/fwk2/graphs_scheduling/src/main.jl:92 [inlined]
 [11] open(::var"#17#18"{Vector{TimespanLogging.Timespan}}, ::String, ::Vararg{String}; kwargs::@Kwargs{})
    @ Base ./io.jl:396
 [12] open(::Function, ::String, ::String)
    @ Base ./io.jl:393
 [13] main(graphs_map::Dict{String, String})
    @ Main ~/fwk2/graphs_scheduling/src/main.jl:91
 [14] top-level scope
    @ ~/fwk2/graphs_scheduling/src/main.jl:106
in expression starting at /home/mafila/fwk2/graphs_scheduling/src/main.jl:106
┌ Warning: Worker 5 died, rescheduling work
└ @ Dagger.Sch ~/.julia/packages/Dagger/kBlIi/src/sch/Sch.jl:545
┌ Error: Error assigning workers
│   exception =
│    ProcessExitedException(5)
│    Stacktrace:
│     [1] worker_from_id(pg::Distributed.ProcessGroup, i::Int64)
│       @ Distributed ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Distributed/src/cluster.jl:1093
│     [2] worker_from_id
│       @ ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Distributed/src/cluster.jl:1090 [inlined]
│     [3] remote_do
│       @ ~/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/share/julia/stdlib/v1.10/Distributed/src/remotecall.jl:557 [inlined]
│     [4] cleanup_proc(state::Dagger.Sch.ComputeState, p::OSProc, log_sink::TimespanLogging.LocalEventLog)
│       @ Dagger.Sch ~/.julia/packages/Dagger/kBlIi/src/sch/Sch.jl:408
│     [5] monitor_procs_changed!(ctx::Context, state::Dagger.Sch.ComputeState)
│       @ Dagger.Sch ~/.julia/packages/Dagger/kBlIi/src/sch/Sch.jl:890
│     [6] (::Dagger.Sch.var"#100#102"{Context, Dagger.Sch.ComputeState})()
│       @ Dagger.Sch ~/.julia/packages/Dagger/kBlIi/src/sch/Sch.jl:508
└ @ Dagger.Sch ~/.julia/packages/Dagger/kBlIi/src/sch/Sch.jl:510
```
Then it reports a series a warnings about workers dying and rescheduling - but it's a known problem

**Comment by SmalRat** [comment]
> Running `graphs_scheduling/src/main.jl` gives the following error once all the processors are done:
> 
> ```julia
> ERROR: LoadError: KeyError: key 68 not found
> Stacktrace:
>   [1] getindex
>     @ ./dict.jl:498 [inlined]
>   [2] _proc_color(ctx::@NamedTuple{proc_to_color::Dict{Dagger.Processor, String}, proc_colors::Vector{RGB{FixedPointNumbers.N0f8}}, proc_color_idx::Base.RefValue{Int64}, proc_to_shape::Dict{Type, String}, proc_shapes::Tuple{String, String, String}, proc_shape_idx::Base.RefValue{Int64}, id_to_proc::Dict{Int64, Dagger.Processor}}, id::Int64)
>     @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:176
>   [3] write_edge(io::IOStream, ts_move::TimespanLogging.Timespan, logs::Vector{TimespanLogging.Timespan}, ctx::@NamedTuple{proc_to_color::Dict{Dagger.Processor, String}, proc_colors::Vector{RGB{FixedPointNumbers.N0f8}}, proc_color_idx::Base.RefValue{Int64}, proc_to_shape::Dict{Type, String}, proc_shapes::Tuple{String, String, String}, proc_shape_idx::Base.RefValue{Int64}, id_to_proc::Dict{Int64, Dagger.Processor}}, inputname::String, inputarg::RemoteChannel{Channel{Int64}})
>     @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:235
>   [4] write_dag(io::IOStream, logs::Vector{TimespanLogging.Timespan}, t::Nothing)
>     @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:323
>   [5] write_dag
>     @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:260 [inlined]
>   [6] _show_plan
>     @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:353 [inlined]
>   [7] show_logs
>     @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:369 [inlined]
>   [8] show_logs(io::IOStream, logs::Vector{TimespanLogging.Timespan}, vizmode::Symbol; options::@Kwargs{})
>     @ Main.ModGraphVizSimpleExt ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:363
>   [9] show_logs
>     @ ~/fwk2/dagger_exts/GraphVizSimpleExt.jl:363 [inlined]
>  [10] #17
>     @ ~/fwk2/graphs_scheduling/src/main.jl:92 [inlined]
>  [11] open(::var"#17#18"{Vector{TimespanLogging.Timespan}}, ::String, ::Vararg{String}; kwargs::@Kwargs{})
>     @ Base ./io.jl:396
>  [12] open(::Function, ::String, ::String)
>     @ Base ./io.jl:393
>  [13] main(graphs_map::Dict{String, String})
>     @ Main ~/fwk2/graphs_scheduling/src/main.jl:91
>  [14] top-level scope
>     @ ~/fwk2/graphs_scheduling/src/main.jl:106
> in expression starting at /home/mafila/fwk2/graphs_scheduling/src/main.jl:106
> ```
> 
> Then it reports a series a warnings about workers dying and rescheduling - but it's a known problem

That is a bug I mentioned recently. The error is thrown when the logs are plotted before all the nodes finish computations, and it happens most of the time. I am currently working on fixing this problem.

**Review by m-fila — COMMENTED** [review]
> That is a bug I mentioned recently. The error is thrown when the logs are plotted before all the nodes finish computations, and it happens most of the time. I am currently working on fixing this problem.

I think I found the culprit

The problem appears because the logs are fetched before all the tasks are finished. For regular tasks their results are fetched when plotting individual graphs, but we never wait for the completion of the extra tasks that sends notifications.
We could get the tasks for that extra node which also happens to be the last node in a graph and wait for their completion

**Comment by SmalRat** [comment]
> > That is a bug I mentioned recently. The error is thrown when the logs are plotted before all the nodes finish computations, and it happens most of the time. I am currently working on fixing this problem.
> 
> I think I found the culprit
> 
> The problem appears because the logs are fetched before all the tasks are finished. For regular tasks their results are fetched when plotting individual graphs, but we never wait for the completion of the extra tasks that sends notifications. We could get the tasks for that extra node which also happens to be the last node in a graph and wait for their completion

Oh, sorry, I did know the exact problem source, just did not describe it much, as was already working on the fix. It started to look too cluttered and hard to read for me, so I added some wrappers etc, consequently, it took a bit more time than expected) Besides, it should be easier to track which tasks are being run and which were already done by now. Moreover, now graphs are parsed one at a time.
However, if it is simpler to stay with the old style of code, I can commit your changes

File: graphs_scheduling/src/main.jl [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    push!(running_dags, uuid)
+    println("Dispatcher: graph scheduled - $uuid: $(get_name(tracked_task_dag))")
+    Threads.@spawn begin
+        wait(tracked_task_dag)
+        put!(notifications, string(get_uuid(tracked_task_dag)))
+    end
+end
+
+function wait_dags_to_finish(notifications::Channel{String}, dags::Dict{String, TrackedTaskDAG}, running_dags::Set{String},
+     done_dags::Set{String}, num=length(waiting_notifiers))
```
Comment (m-fila) [inline graphs_scheduling/src/main.jl:97]: `waiting_notifiers` is not defined

Comment (SmalRat) [reply graphs_scheduling/src/main.jl]: Fixed, that was after the latest "cleanup"

**Comment by SmalRat** [comment]
> I think I'd rather take the version before adding `AbstractMetaTask` end the others. Refactoring of scheduling graph pipeline should go to a different PR where we could focus solely on it

OK, I will try to revert these changes

**Comment by m-fila** [comment]
> OK, I will try to revert these changes

I have a different proposal. Let's freeze this as is and I'll cherry pick the relevant commits to a new PR with a clean history and no conflicts

**Comment by SmalRat** [comment]
> > OK, I will try to revert these changes
> 
> I have a different proposal. Let's freeze this as is and I'll cherry pick the relevant commits to a new PR with a clean history and no conflicts

Oh, sorry, I've seen this too late
I have created a separate branch for scheduling pipeline refactoring and reverted these changes here to the state of [8a52e6d](https://github.com/key4hep/key4hep-julia-fwk/pull/7/commits/8a52e6d8a9cf31571a7d03be2af3ffe39ff592be)

**Comment by SmalRat** [comment]
> > > OK, I will try to revert these changes
> > 
> > 
> > I have a different proposal. Let's freeze this as is and I'll cherry pick the relevant commits to a new PR with a clean history and no conflicts
> 
> Oh, sorry, I've seen this too late I have created a separate branch for scheduling pipeline refactoring and reverted these changes here to the state of [8a52e6d](https://github.com/key4hep/key4hep-julia-fwk/pull/7/commits/8a52e6d8a9cf31571a7d03be2af3ffe39ff592be)

Now I can apply your suggestions on silencing the workers and a bug fix. But will wait for the response this time :)

**Comment by m-fila** [comment]
No problem. Please take a look at #7 and comment if anything is missing or I messed up something :sweat_smile:

**Comment by SmalRat** [comment]
> No problem. Please take a look at #11 and comment if anything is missing or I messed up something 😅

Looks like everything is fine there

**Pull request closed without merging**